src/thor/trace_route_action.cc

```diff
@@ -419,34 +423,61 @@ thor_worker_t::map_match(Api& request) {
+          edge_table[path_edge_itr->edgeid] = ++idx;
+        }
       }
 
       // here we mark the original shape points that just after and just before the
       // discontinuities so that in the path they get "upgraded" to break points
       auto match_result_itr = match_results.cbegin();
       for (auto& edge_group : edge_groups) {
-        // we know the discontinuity happens on this edge so we just need to
-        // find the first match result on this edge
-        auto first_edge_id = std::get<0>(edge_group)->edgeid;
-        while (match_result_itr != match_results.cend() &&
-               match_result_itr->edgeid != first_edge_id) {
-          ++match_result_itr;
-        }
+        auto first_edge = std::get<0>(edge_group);
+        auto last_edge = std::get<1>(edge_group);
+
+        match_result_itr = std::find_if(match_result_itr, match_results.cend(),
+                                        [first_edge](const MatchResult& result) {
+                                          return result.edgeid == first_edge->edgeid;
+                                        });
         std::get<2>(edge_group) = match_result_itr;
 
-        // we know the discontinuity happens on this edge so we just need to
-        // find the last match result on this edge
-        auto last_edge_id = std::get<1>(edge_group)->edgeid;
-        while (match_result_itr != match_results.cend() && match_result_itr->edgeid != last_edge_id) {
-          ++match_result_itr;
-        }
-        while (match_result_itr != match_results.cend() && match_result_itr->edgeid == last_edge_id) {
+        // TODO:: handle case when both discontinuity and loop on last edge
+        // we know the discontinuity happens on this edge so we need to
+        // find the first match result on this edge
+        match_result_itr = std::find_if(std::next(match_result_itr), match_results.cend(),
+                                        [last_edge](const MatchResult& result) {
+                                          return result.edgeid == last_edge->edgeid;
+                                        });
+
+        bool has_loop_on_last_edge = std::get<4>(edge_group);
+        // find the last match result on this edge before discontinuity
+        while (match_result_itr != match_results.end() &&
+               match_result_itr->edgeid == last_edge->edgeid) {
           ++match_result_itr;
+          // this indicates discontinuity on the same edge (if there is no loop) see below:
+          //
+          //     curr distance_along    prev distance_along
+          //            |                       |
+          //            ▼                       ▼
+          //  X---------------------------------------------> 100%
+          //
+          if (std::prev(match_result_itr)->edgeid == last_edge->edgeid &&
+              match_result_itr->distance_along < std::prev(match_result_itr)->distance_along &&
+              !has_loop_on_last_edge) {
```

Comment:
so here we catch the case where there was a discontinuity on the same edge from the `FormPath` function above. but we also have to take care that there wasnt a loop that lead us to seeing the same edge twice but with different distance along in terms of order. i cant help but think there is an easier way to do this. like cant we just note when we have the same edgeid twice in a row in the `EdgeSegment` vector? in other words we can chop up our edge groups based on the edge segments vector as well and not have to know about the loops at all. i havent looked at this code in a while so maybe its not possible. just thinking out loud